File: Semestrul 2/OOP/Examen OOP/EXAMEN PRACTIC/Tractoare/Tractoare/Observable.cpp
#include "Observable.h"

#include <algorithm>

void Observable::add_observer(Observer* obs)
{
	std::vector<Observer*>::iterator it = find(observers.begin(), observers.end(), obs);
	if (it == observers.end())
		observers.push_back(obs);
}

void Observable::del_observer(Observer* obs)
{
	std::vector<Observer*>::iterator it = find(observers.begin(), observers.end(), obs);
	if (it != observers.end())
		observers.erase(it);
}

void Observable::notify_observers()
{
	for (auto& i : observers)
		i->update_me();
}

File: Semestrul 4/MPP/Java/HolidaysAppSemester/src/main/java/com/example/holidaysappsemester/utils/observer/Observable.java
package com.example.holidaysappsemester.utils.observer;


import com.example.holidaysappsemester.utils.events.Event;

public interface Observable<E extends Event> {
    void addObserver(Observer<E> e);
    void removeObserver(Observer<E> e);
    void notifyObservers(E t);
}
